clang/lib/CodeGen/MicrosoftCXXABI.cpp

```diff
@@ -1105,6 +1105,11 @@ bool MicrosoftCXXABI::hasMostDerivedReturn(GlobalDecl GD) const {
 
 static bool isTrivialForMSVC(const CXXRecordDecl *RD, QualType Ty,
                              CodeGenModule &CGM) {
+  // If the record is marked with the trivial_abi attribute, we don't
+  // have to conform to the standard MSVC ABI.
+  if (RD->hasAttr<TrivialABIAttr>())
```

Comment:
I've realized this misses a case, consider this a wrapper struct that contains a trivial_abi subobject: https://godbolt.org/z/bK6aGv4d7

I spent 15 minutes looking through the things we track for trivial_abi, and unfortunately, we don't track this anywhere already, so the best solution I can come up with is to do a recursive walk of all the subobjects and look for the attribute. :(

Reply:
That's not how trivial_abi normally works?  You'll never even reach this code for anything with a non-trivial destructor.

I guess if you're abusing trivial_abi just to get the MSVC-specific behavior in this patch, you might want that recursive behavior, but that seems like it's stretching it a bit.  We might want to consider a dedicated attribute.

------

On a related note, if we're going to mess with this anyway, we might want to mess with the canPassInRegisters() in clang/lib/Sema/SemaDeclCXX.cpp also, so structs with a deleted copy constructor work, for example:

```
struct __attribute__((trivial_abi)) Trivial {
  int *p = nullptr;
  Trivial(const Trivial &) noexcept = delete;
  Trivial(Trivial &&) noexcept;
  ~Trivial();
};
Trivial f(Trivial x) { return x;}
```

Reply:
> That's not how trivial_abi normally works?

I think it is how it works, based on re-reading the code just now and the test case I linked.

To your point, the MSVC-specific code in `canPassInRegisters` is full of bugs, but I don't want to ask @tru  to deal with that.

Reply:
Oh, I see, whether the constructors/destructor/etc. are trivial for calls is determined recursively; we don't record whether that was because of an explicit trivial_abi attribute, or because there just weren't any non-trivial fields.  So we can't tell the difference here.

It feels sort of weird to treat a struct with a trivial_abi member as somehow more trivial than a struct that just contains plain data.

Reply:
So if I understand correctly:

- There are additional places in the microsoft code path where we don't handle trivial cases without the attribute.
- The glaring issue here with this PR is that if you put a trivial_abi struct in a struct it won't be marked as trivial itself since it doesn't search recursively.

Seems like there are many more places where we can do better, do we want to block this PR on the recursive search? What would the rules for that be? It can't just be that if you have a struct with a trivial_abi attribute inside any struct the parent should also be considered trivial. Would it be that isTrivialForMicrosoft would return true if the current struct passes the checks there and only contains a struct with a trivial_abi attribute?

Reply:
Yes, I think I do want to block this on the recursive search. Even if it's ugly and we ultimately replace it with an explicitly tracked and type trait (`containsTrivialAbiSubobject`?), it reduces the number of externally visible ABI changes. I know we're saying this is ABI unstable already, but I still think fewer breaks is better.

Reply:
Happy to add this. Can you point me to an example where we search recursively or give some pseudo code example and I should be able to take it from there pretty soon.

Reply:
The pseudo code version is something like:
  bool containsTrivialAbiSubobject(CXXRecordDecl *RD) {
    if (RD->hasAttr<TrivialABIAttr>()) return true;
    for (auto B : RD->bases()) {
      if (containsTrivialAbiSubobject(B->getAsCXXRecordDecl()))
        return true;
    }
    for (auto F : RD->fields()) {
      const CXXRecordDecl* FRD = F->getType()->getAsCXXRecordDecl();
      if (FRD && containsTrivialAbiSubobject(FRD))
        return true;
    }    
    return false;
  }

A bit of a gross inefficient DFS, with some CXXBaseSpecifier details omitted. It might also be worth tracking Visited records to handle fields of the same record type without recursing over all subobjects again.